external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java

```diff
@@ -0,0 +1,654 @@
+        throw new RuntimeException("Unable to create lock directory: " + lockDirPath);
+      }
+    } catch (IOException e) {
+      LOG.error("Unable to create lock dir: " + lockDirPath, e);
+      throw new RuntimeException(e.getMessage(), e);
+    }
+
+    // -- lock timeout
+    if( conf.get(Configs.LOCK_TIMEOUT) !=null )
+      this.lockTimeoutSec =  Integer.parseInt(conf.get(Configs.LOCK_TIMEOUT).toString());
+
+    // -- enable/disable ACKing
+    Object ackers = conf.get(Config.TOPOLOGY_ACKER_EXECUTORS);
+    if( ackers!=null )
+      this.ackEnabled = ( Integer.parseInt( ackers.toString() ) > 0 );
+    else
+      this.ackEnabled = false;
+
+    // -- commit frequency - count
+    if( conf.get(Configs.COMMIT_FREQ_COUNT) != null )
+      commitFrequencyCount = Integer.parseInt( conf.get(Configs.COMMIT_FREQ_COUNT).toString() );
+
+    // -- commit frequency - seconds
+    if( conf.get(Configs.COMMIT_FREQ_SEC) != null )
+      commitFrequencySec = Integer.parseInt( conf.get(Configs.COMMIT_FREQ_SEC).toString() );
+
+    // -- max duplicate
+    if( conf.get(Configs.MAX_DUPLICATE) !=null )
+      maxDuplicates = Integer.parseInt( conf.get(Configs.MAX_DUPLICATE).toString() );
+
+    // -- clocks in sync
+    if( conf.get(Configs.CLOCKS_INSYNC) !=null )
+      clocksInSync = Boolean.parseBoolean(conf.get(Configs.CLOCKS_INSYNC).toString());
+
+    // -- spout id
+    spoutId = context.getThisComponentId();
+
+    // setup timer for commit elapse time tracking
+    setupCommitElapseTimer();
+  }
+
+  private String getDefaultLockDir(Path sourceDirPath) {
+    return sourceDirPath.toString() + Path.SEPARATOR + Configs.DEFAULT_LOCK_DIR;
+  }
+
+  private static void checkValidReader(String readerType) {
+    if(readerType.equalsIgnoreCase(Configs.TEXT)  || readerType.equalsIgnoreCase(Configs.SEQ) )
+      return;
+    try {
+      Class<?> classType = Class.forName(readerType);
+      classType.getConstructor(FileSystem.class, Path.class, Map.class);
+      return;
+    } catch (ClassNotFoundException e) {
+      LOG.error(readerType + " not found in classpath.", e);
+      throw new IllegalArgumentException(readerType + " not found in classpath.", e);
+    } catch (NoSuchMethodException e) {
+      LOG.error(readerType + " is missing the expected constructor for Readers.", e);
+      throw new IllegalArgumentException(readerType + " is missing the expected constuctor for Readers.");
+    }
+  }
+
+  @Override
+  public void ack(Object msgId) {
+    MessageId id = (MessageId) msgId;
+    inflight.remove(id);
+    ++acksSinceLastCommit;
+    tracker.recordAckedOffset(id.offset);
+    commitProgress(tracker.getCommitPosition());
+    if(fileReadCompletely) {
+      markFileAsDone(reader.getFilePath());
+      reader = null;
+    }
+    super.ack(msgId);
+  }
+
+  private boolean canCommitNow() {
+    if( acksSinceLastCommit >= commitFrequencyCount )
+      return true;
+    return commitTimeElapsed.get();
```

Comment:
Th timer itself is not required since the commit is performed synchronously (whenever nextTuple, ack is invoked). I don't see the commits being triggered asynchronously from the timer.